### PR TITLE
Added a check  in MetaApiController to check table exist in bigquery

### PR DESCRIPTION
**Pull request by dionboles-asym**
status check #36

### CODE REVIEW

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 1]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
...
+    try {
+      BigQuery bigQuery = BigQueryOptions.newBuilder().setProjectId("gdc-bq-sample").build().getService();
+      String StatusCheck = bigQuery.getDataset("cda_mvp").getDatasetId().getDataset();
+      if (StatusCheck.equals("cda_mvp")) {
+        SystemStatusSystems otherSystemStatus = new SystemStatusSystems().ok(true).addMessagesItem("everything is fine");
+        systemStatus.ok(true).putSystemsItem("BigQuery", otherSystemStatus);
+        return new ResponseEntity<>(systemStatus, HttpStatus.OK);
+      }
+    } catch (Exception e) {
+      logger.error(e.getMessage());
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:49]: When logging exceptions, you almost always want to log the exception itself, so the log will show the stack trace from the exception.
```suggestion
      logger.error("status check failed", e);
```

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 2]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
...
+      SystemStatusSystems otherSystemStatus = new SystemStatusSystems();
+      otherSystemStatus.setOk(false);
+      otherSystemStatus.addMessagesItem("Error");
+      systemStatus = new SystemStatus().putSystemsItem("BigQuery", otherSystemStatus);
+      logger.error(e.toString());
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:54]: This doesn't seem necessary, why is this needed?

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 3]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
...
+      BigQuery bigQuery = BigQueryOptions.newBuilder().setProjectId("gdc-bq-sample").build().getService();
+      String StatusCheck = bigQuery.getDataset("cda_mvp").getDatasetId().getDataset();
+      if (StatusCheck.equals("cda_mvp")) {
+        SystemStatusSystems otherSystemStatus = new SystemStatusSystems().ok(true).addMessagesItem("everything is fine");
+        systemStatus.ok(true).putSystemsItem("BigQuery", otherSystemStatus);
+        return new ResponseEntity<>(systemStatus, HttpStatus.OK);
+      }
+    } catch (Exception e) {
+      logger.error(e.getMessage());
+      SystemStatusSystems otherSystemStatus = new SystemStatusSystems();
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:50]: There's no need to have two different `SystemStatusSystems` objects. If you create this outside the `try` and call it something like `bigQuerySystemStatus`, you can create it once and update it in both the `try` and `catch` blocks.

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 4]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
...
   public ResponseEntity<SystemStatus> serviceStatus() {
-    HttpStatus httpStatus = HttpStatus.OK;
-
-    SystemStatusSystems otherSystemStatus =
-        new SystemStatusSystems().ok(true).addMessagesItem("everything is fine");
+    SystemStatus systemStatus = new SystemStatus();
+    try {
+      BigQuery bigQuery = BigQueryOptions.newBuilder().setProjectId("gdc-bq-sample").build().getService();
+      String StatusCheck = bigQuery.getDataset("cda_mvp").getDatasetId().getDataset();
+      if (StatusCheck.equals("cda_mvp")) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:43]: The logic here isn't correct. If the BQ call doesn't throw an exception but this check fails, the status will be empty.

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 5]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
 
   @Override
   public ResponseEntity<SystemStatus> serviceStatus() {
-    HttpStatus httpStatus = HttpStatus.OK;
-
-    SystemStatusSystems otherSystemStatus =
-        new SystemStatusSystems().ok(true).addMessagesItem("everything is fine");
+    SystemStatus systemStatus = new SystemStatus();
+    try {
+      BigQuery bigQuery = BigQueryOptions.newBuilder().setProjectId("gdc-bq-sample").build().getService();
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:41]: Duplicating the code in `QueryService` here can result in an incorrect test. Instead this code should call a method in `QueryService` that returns the BQ status, by using the bigquery object in that class.

File: src/main/java/bio/terra/cda/app/controller/MetaApiController.java [hunk 6]
```diff
@@ -30,16 +36,27 @@ public MetaApiController(ApplicationConfiguration applicationConfiguration) {
...
+      }
+    } catch (Exception e) {
+      logger.error(e.getMessage());
+      SystemStatusSystems otherSystemStatus = new SystemStatusSystems();
+      otherSystemStatus.setOk(false);
+      otherSystemStatus.addMessagesItem("Error");
+      systemStatus = new SystemStatus().putSystemsItem("BigQuery", otherSystemStatus);
+      logger.error(e.toString());
+    }
+    return new ResponseEntity<>(systemStatus, HttpStatus.BAD_REQUEST);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/cda/app/controller/MetaApiController.java:56]: I believe the proper HTTP status code for a status failure is 503, `SERVICE_UNAVAILABLE`

**Review by pshapiro4broad — COMMENTED** [review]
The basic idea is OK, although I have a few comments about the implementation.

Also, the bigger issue here is that we don't want to call BQ every time the status endpoint is called, as the GKE cluster management software can call it very frequently. As described in https://github.com/CancerDataAggregator/cda-service/issues/36, what's needed here is a scheduled task or cached result. I would try using a spring scheduled task do this.  https://spring.io/guides/gs/scheduling-tasks/

File: src/main/java/bio/terra/cda/app/service/QueryService.java [hunk 7]
```diff
@@ -46,7 +52,34 @@
...
+    }catch (Exception e){
+      logger.error("Status check failed ", e);
+    }
+    if(success){
+      bigQuerySystemStatus.ok(true).addMessagesItem("everything is fine");
+    }else{
 
+      bigQuerySystemStatus.ok(false).addMessagesItem("Error");
+    }
+    systemStatus.ok(bigQuerySystemStatus.getOk()).putSystemsItem("BigQuery",bigQuerySystemStatus);
```
Comment (fkaufman-asym) [inline src/main/java/bio/terra/cda/app/service/QueryService.java:80]:
```suggestion
    systemStatus.ok(bigQuerySystemStatus.getOk()).putSystemsItem("BigQueryStatus",bigQuerySystemStatus);
```

Comment (fkaufman-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:80]: I don't know what this should be,  we are trying to indicate that the DataStore is reachable.  BigQuery:'ok' sounds like Google's not broken.   We may need this hook to marinate to determine what is the additional information needed in this return.

File: src/main/java/bio/terra/cda/app/service/BigQueryClearCache.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+@Component
+public class BigQueryClearCache {
+    private static final Logger logger = LoggerFactory.getLogger(BigQueryClearCache.class);
+
+    @Autowired
+    QueryService queryService;
+
+    @Scheduled(fixedRate = 120000) // 2 min refresh
+    public void task() {
+        logger.info("Scheduler has updated cache");
```
Comment (fkaufman-asym) [inline src/main/java/bio/terra/cda/app/service/BigQueryClearCache.java:19]: This is probably a debug statement
```suggestion
        logger.debug("Scheduler has updated cache");
```

**Review by fkaufman-asym — CHANGES_REQUESTED** [review]

**Comment by fkaufman-asym** [comment]
I had to run the spotlessApply flag when building to correct the code formatting.  You may need to use that flag before checking in the next time to pass the linting checks.

File: src/main/java/bio/terra/cda/app/service/QueryService.java (hunk 7 above)
Comment (dionboles-asym) [reply src/main/java/bio/terra/cda/app/service/QueryService.java:80]: That is build into the systemStatus class the ok method dose a boolen check and sets the value of ok

**Review by nemanjab17 — APPROVED** [review]

**Review by fkaufman-asym — APPROVED** [review]
I checked this out and it seems to be operational as coded.
